### ISSUE TITLE
AssertionError when first typing in a fresh shell

### ISSUE THREAD

**Issue by qwenger** [xonsh/xonsh#3586]
This is a follow-up on #2339.

Since the upgrade to PTK3, I sometimes (like 1/5 times) get the traceback reproduced below when typing in a freshly opened terminal.

Pressing ENTER reprints the same traceback indefinitely.

I'm on ArchLinux, using xfce4-terminal; but it also happens in linux console (CTRL+ALT+Fx).

It can happen even when waiting for several seconds before typing, and happens after a couple characters are entered.

## xonfig

<details>

```
+------------------+---------------------+
| xonsh            | 0.9.18              |
| Git SHA          | 26475146            |
| Commit Date      | May 5 06:42:28 2020 |
| Python           | 3.8.2               |
| PLY              | 3.11                |
| have readline    | True                |
| prompt toolkit   | 3.0.5               |
| shell type       | prompt_toolkit      |
| pygments         | 2.6.1               |
| on posix         | True                |
| on linux         | True                |
| distro           | arch                |
| on darwin        | False               |
| on windows       | False               |
| on cygwin        | False               |
| on msys2         | False               |
| is superuser     | False               |
| default encoding | utf-8               |
| xonsh encoding   | utf-8               |
| encoding errors  | surrogateescape     |
+------------------+---------------------+
```
</details>

### Traceback (if applicable)

<details>

```
Unhandled exception in event loop:
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/eventloop/utils.py", line 72, in schedule
    func()
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 456, in redraw
    self._redraw()
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 523, in _redraw
    self.context.run(run_in_context)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 509, in run_in_context
    self.renderer.render(self, self.layout)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/renderer.py", line 598, in render
    layout.container.write_to_screen(
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 381, in write_to_screen
    sizes = self._divide_heights(write_position)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 435, in _divide_heights
    dimensions = [c.preferred_height(width, height) for c in self._all_children]
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 435, in <listcomp>
    dimensions = [c.preferred_height(width, height) for c in self._all_children]
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 792, in preferred_height
    return self.content.preferred_height(width, max_available_height)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 324, in preferred_height
    dimensions = [
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 325, in <listcomp>
    c.preferred_height(width, max_available_height) for c in self._all_children
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 2595, in preferred_height
    return self.content.preferred_height(width, max_available_height)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 1620, in preferred_height
    return self._merge_dimensions(
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 1648, in _merge_dimensions
    preferred = get_preferred()
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 1613, in preferred_content_height
    return self.content.preferred_height(
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/controls.py", line 639, in preferred_height
    content = self.create_content(width, height=1)  # Pass a dummy '1' as height.
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/controls.py", line 761, in create_content
    document = buffer.document
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/buffer.py", line 524, in document
    return self._document_cache[
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/cache.py", line 98, in __missing__
    result = self.get_value(*key)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/document.py", line 98, in __init__
    assert cursor_position is None or cursor_position <= len(text), AssertionError(

Exception cursor_position=2, len_text=1
Press ENTER to continue... 
```
</details>

**Comment by bobhy** [comment]
There weren't any ptk3-specific code changes in that update, mostly a jettisoning of legacy ptk1 support.  So... did you make any changes in `.xonshrc` when you started running ptk3?  Could you share your `.xonshrc` as well?

**Comment by qwenger** [comment]
I'm using the default history backend, that is files.

Could you please explain in more details what you mean by "doubling or redoubling" my history? Thanks!

**Comment by bobhy** [comment]
I'm looking for ways to make your environment a more reliable repro for this error, was thinking you could make additional copies of your history files so Xonsh takes longer to load them and remains in the crash window longer. But there must be some other variable, too.  Would you post your .xonshrc as well?  

The other thread has some good ideas for what might be happening, I'll try to come up with a fix but it would be good to have a reliable test for it.

**Comment by qwenger** [comment]
Ok, I'll try. Here's my .xonshrc:

```
import xonsh

#
# ~/.xonshrc
#

$SHELL_TYPE = "prompt_toolkit"

# https://github.com/xonsh/xonsh/issues/3579
# For now, just source-bash what's needed.
for fn in g`/etc/profile.d/*.sh`:
    if "gpm" not in fn:
        source-bash @(fn)

def env_branch():
    e = xonsh.prompt.env.env_name() or ""
    b = xonsh.prompt.vc.current_branch() or ""
    c = xonsh.prompt.vc.branch_color()
    # XXX: xonsh does not handle f-strings correctly...
    ce = "{NO_COLOR}"
    return f"({e}|{c}{b}{ce}) " if e or b else ""

$PROMPT_FIELDS["env_prefix"] = ""
$PROMPT_FIELDS["env_postfix"] = ""
$PROMPT_FIELDS["env_branch"] = env_branch
$PROMPT = lambda: "{env_branch}{BOLD_YELLOW}{user}@{hostname}{NO_COLOR}:{BOLD_BLUE}{cwd}\n{BOLD_RED}${NO_COLOR} "

$COMPLETIONS_CONFIRM = True
$COMPLETIONS_DISPLAY = "readline"
$XONSH_COLOR_STYLE = "default"
$XONSH_SHOW_TRACEBACK = False
$SUGGEST_COMMANDS = False

xontrib load vox

$PROJECT_DIRS = [
    "/data/dev/python",
    "/data/dev/c_cpp",
]

import os
$VIRTUALENV_HOME = os.path.expanduser("~/.local/share/virtualenvs")

xontrib load avox

# https://github.com/michael-lazar/rtv/issues/614
from xonsh.commands_cache import predict_false
__xonsh__.commands_cache.threadable_predictors["git"] = predict_false
```

**Comment by anki-code** [comment]
> $PROMPT = lambda: "{env_branch}{BOLD_YELLOW}{user}@{hostname}{NO_COLOR}:{BOLD_BLUE}{cwd}\n{BOLD_RED}${NO_COLOR} "

Try https://github.com/anki-code/xontrib-prompt-bar :)

**Comment by qwenger** [comment]
@bobhy so I have roughly 9000 history files. I wrote a small script to make a bunch of copies (with a carefully-crafted uuid in the name so that I can delete them easily afterwards) of the largest file (~390kB).

With 10000 additional copies, the prompt is barely usable and tab-completion mostly gives garbled results (something I've seen happening from time to time in my genuine configuration as well)

With 2000 additional copies, the fan goes crazy each time I start typing in a new shell, and I almost cannot finish any command line without the error popping up. Better yet, when by chance I do, the error happens when entering a second command (something I never experienced in my genuine configuration).

Does this meet your expectations?

@anki-code thanks for the suggestion, will give it a try.
...But if I can do it in 9 lines of code without external help, is a xontrib really needed? :-)

**Comment by qwenger** [comment]
Interested in my beautiful script? :D

```
import shutil
import os
b = os.path.expanduser("~/.local/share/xonsh")
heaviest = max(filter(lambda fn: fn.endswith(".json"), os.listdir(b)), key=lambda fn: os.path.getsize(os.path.join(b, fn)))
for i in range(10000):
    shutil.copy(os.path.join(b, heaviest), os.path.join(b, f"xonsh-00000000-0000-4000-0000-000{i:09}.json"))
    print(i)
```

**Comment by bobhy** [comment]
Thanks for the detective work @matpi.  I will hazard a guess that you can
make the error vanish if you're willing to run with a small history.  And I
wonder whether the prompt misbehaves with such a large history when you
revert to ptk2 (maybe?)  or run readline shell (probably not).

**Comment by qwenger** [comment]
> I will hazard a guess that you can make the error vanish if you're willing to run with a small history.

Yes it does vanish (but I'm not really... :-) ).

> And I wonder whether the prompt misbehaves with such a large history when you revert to ptk2 (maybe?)

Nope, even if it is incredibly slow.

> or run readline shell (probably not).

Nope, and the speed is fine (but there's no history completion so that's not fair).

**Comment by bobhy** [comment]
Thanks, @qwenger, for your script.  It gives me a 100% reliable repro, so now I have something to debug with!

**Comment by bobhy** [comment]
I haven't found the smoking gun yet, but limiting the Xonsh history loaded into PTK history to about 8-9000 commands seems to reduce the incidence of crashes. My repro environment was failing 100%, but not once with this fix in tens of tries.  I have a patch for doing this, and you could use Xonsh `history` to access the rest if you needed to go way back.

8128 commands is the default for $XONSH_HISTORY_SIZE, so you'd think we'd rarely see this. 
But Xonsh history garbage collection is not working for the json file backend due to an unrelated issue (#3599),  Fixing this is easy, but it might delete data for those of you with large history.

So one way or the other, there's a way to sweep this bug under the carpet.  

A real fix that resolves all the multitasking hazards between Xonsh and ptk3, is a bigger project.  Discussion in #3435 is relevant here, and does not sound encouraging.

**Comment by qwenger** [comment]
Thank you @bobhy for the update.

Do #3599/#3600 mean that some of my history *.json are actually garbage?

**Comment by bobhy** [comment]
I wouldn't say garbage, but perhaps dead men walking, unless you configure a big retention history in $XONSH_HISTORY_SIZE before #3600 is rolled out.

**Comment by bobhy** [comment]
I think I've found a problem in PTK3 which causes this. Basically, the process of loading history via `ThreadedHistory` is not thread-safe at the point where the history is copied into the `Buffer`.  

Here's a modification to prompt-toolkit: https://github.com/bobhy/python-prompt-toolkit/tree/th-threadsafe-load.  It uses `loop.call_soon_threadsafe()` to update Buffer on the event loop.  Prompt Toolkit in general does not want to assume the event loop is available when history loads, but it works for Xonsh.

@qwenger, can you download this version and see whether it fixes your issue?  I was never able to repro the call stack you originally reported, but using your script, I was able to repro a slightly different failure and this change fixes that one.

Meantime @jonathanslenders, here's a head's up on this situation.  I intend to work up this proposed change as a PR for prompt-toolkit, but not until I have a chance to run regression tests and document this a bit better.  For example, it would be good to have a repro scenario that doesn't depend on Xonsh...

Meantime, 

I was not able to repro the original failure, but

**Comment by qwenger** [comment]
@bobhy I tested your PTK version.

- The error doesn't seem to show up anymore.
- When I add 10000 xonsh history files with my script, many xonsh subprocesses seem to hang at either the startup or the shutdown, my CPU going crazy. For example I type in `htop<enter>`, either it simply doesn't start, or when then quitting it doesn't return to the shell. Trying to list the content of `.local/share/xonsh`, with its about 19000 files, is also a typical example. Sometimes after hanging for a while it displays the following traceback:

```
Exception ignored when trying to write to the signal wakeup fd:
Traceback (most recent call last):
  File "/usr/lib/python3.8/asyncio/unix_events.py", line 42, in _sighandler_noop
    def _sighandler_noop(signum, frame):
BlockingIOError: [Errno 11] Resource temporarily unavailable
```

**Comment by bobhy** [comment]
Good news that the PKT version works for N > 1!  I will submit the PR to PTK team tonight...

The additional hangs and system thrashing are, I think, symptoms of extreme memory pressure.  I believe Xonsh + PTK try to make 3 copies of the entire history in memory! And we clearly are offering a lot of history...  High CPU activity might be frequent (and fruitless) Python garbage collection, and maybe you are also seeing high paging or swap file I/O?

**Comment by laloch** [comment]
The patch does indeed fix the cursor_possition assert, but makes xonsh deadlock and spin endlessly on some futex:
```
...
futex(0x7f9e02f67970, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0x7f9e02f6796c, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=47705, tv_nsec=825471776}, FUTEX_BITSET_MATCH_ANY) = 0
futex(0x7f9e02f67970, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0x7f9e02f67968, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=47705, tv_nsec=826111810}, FUTEX_BITSET_MATCH_ANY) = 0
...
```

**Comment by laloch** [comment]
@bobhy, I suspect you wanted to do something with `BufferHistoryLock` (perhaps a critical section in `new_history_item(...)`?), but you didn't.

**Comment by bobhy** [comment]
The lock is left over from a previous experiment, not missing code from the current approach.  I discovered that locking just in `new_history_item()` simply moved the failure to other places where `self._working_index` is used.  PTK is (proudly!) non-thread-safe, depending instead on event loop synchronization. Path of least resistance was for me to use the event loop too...

Could the seemingly endless spinning be a very big history simply taking many minutes to load?  Do you get a prompt and can you run a command while this is happening?    

Could you detail out a bit more of your repro scenario?  Does it fail reliably?   Would you be willing to share your current history folder?

**Comment by laloch** [comment]
> Could the seemingly endless spinning be a very big history simply taking many minutes to load?

I don't think so. I left it in that state for several minutes and then I attached strace to the running process. It did not show any activity whatsoever apart from the looping over the wait/wake calls on those futexes.

> Do you get a prompt and can you run a command while this is happening?

Nope, no prompt. It even ignores ^C.

> Could you detail out a bit more of your repro scenario? Does it fail reliably?

Well, it happend twice in two attempts...
As far as I can tell I only ran a few commands (like two or three).

> Would you be willing to share your current history folder?

I did already delete it. I used $XONSH_DATA_DIR in a tmpfs.
I can, however, retry tomorrow. I do even have a (sort of) bult-in debugger now in #3606.

**Comment by bobhy** [comment]
@laloch, No need to work on the retry, I was able to repro approximately similar behavior and found something to fix.  This one's on Xonsh.  

Here is my tale of woe so far.  @jonathanslenders, this involves ptk3 as well...

0. Kudos to @qwenger for providing a script to create an outrageously large Xonsh history on disk, 10k files, 3M commands.  Without that tool, I would never have been able to figure this out.
1. Xonsh has a bug that does not actually prune history for json backend (#3599). This allows senior users to accumulate large histories over time.  (Large > 10k to 100k commands, in my experience).  This creates the opportunity to observe the following behaviors...
2. PTK3 has an interface that loads all history from its caller into memory, but `Buffer` doesn't want to modify that, so it makes a second copy into working storage.  Xonsh uses the `ThreadedHistory` interface to load history on a background thread, so user can get a prompt immediately, with whatever history that has been loaded so far.   
3. It takes minutes to load 10K files, plenty of time for user to try multiple commands.  I'm not quite sure why, but the history loading process gets slower as it grinds on, O(1) behavior to a first approximation.  I'm guessing it is simply memory pressure, but it could be GIL contention or I know not what.  But I observe that whatever it is, it makes all threads run slow, and this is the behavior that opens up many timing windows that would not "normally" happen.  Here's one example, from `xonsh.history.json.JsonHistory`:

```python
def all_items(self, newest_first=False, **kwargs):
        """
        Returns all history as found in XONSH_DATA_DIR.

        yield format: {'inp': cmd, 'rtn': 0, ...}
        """
        while self.gc and self.gc.is_alive():
            time.sleep(0.011)  # gc sleeps for 0.01 secs, sleep a beat longer
        for f in _xhj_get_history_files(newest_first=newest_first):
            try:
                json_file = xlj.LazyJSON(f, reopen=False)
            except ValueError:
    . . .
```
In my testing, the `xlj.LazyJSON()` failed (after histry had been loading for a while) with "file not found", because the history loading thread enumerated files when first started, but there was plenty of time for me to issue a `history gc` command and delete some of those files before it got to them.

4. Your Xonsh has an additional problem that makes things run much slower in this scenario. I think this is the culprit causing @laloch's observation of CPU going crazy: in `xonsh.history.json.JsonHistory`, there's this:

```python
def run_gc(self, size=None, blocking=True):
        self.gc = JsonHistoryGC(wait_for_shell=False, size=size)
        if blocking:
            while self.gc.is_alive():
                continue
```
I replaced the `continue` with a `time.sleep(0.1)` that that frees up a lot of CPU and makes it possible to tolerate loading of even more of this unreasonably large history!

The strace might be normal behavior of the `while True` loop in `xonsh.proc.populate_fd_queue()` waiting for whatever command you issued to complete.  In any case, it's on my stack when I repro the problem, YMMV. 

### What to do about all this?

1. Fix the timing hazard in PTK ThreadedHistory.  I feel pretty sure it was responsible for the initial assert fail that kicked off this issue.  I've got a [draft fix](https://github.com/bobhy/python-prompt-toolkit/tree/th-threadsafe-load), but need to open an issue and discuss with ptk team.  Need a simple repro for the issue..
2. Fix the no GC bug (#3599), and add the fix above.  I've got a PR ready to go for this, it was in the course of testing it with the unreasonably large history that I duplicated the "deadlock" behavior (which I think was "merely sleeping" behavior).
3. Redesign Xonsh history, paying more attention to multitasking.  I think this is an opportunity to start in with `asyncio` and share the event loop with the ptk prompt.  I'll open a separate issue for item 3. above and we can discuss it there.

**Comment by bobhy** [comment]
@laloch, @jonathanslenders, New issue posted in prompt-toolkit/python-prompt-toolkit#1158, with clean repro along with PR for draft of fix.

**Comment by bhawkins** [comment]
Just wanted to report that I've been getting this issue on debian unstable with xonsh 0.9.18 and PTK 3.0.5 packages from the repos.  I do have a large history, which I find very useful:

```
$ grep HISTORY ~/.xonshrc                                                       
$XONSH_HISTORY_SIZE = (2, 'years')
$ ls ~/.local/share/xonsh/*json | wc                                            
   3836    3836  314552                             
$ cat ~/.local/share/xonsh/*json | wc                                           
 164536 4385216 47504442
```
I get a crash almost every time I open a new terminal unless I remember to pause a moment before I start typing. I didn't know about the sqlite history backend. Is it immune? If so, is there a way to convert existing history from JSON to sqlite? Or maybe a way to prune my history to something reasonable?

**Comment by bobhy** [comment]
If you're feeling adventurous, you could give this a try: prompt-toolkit/python-prompt-toolkit#1170.  It fixes a PTK issue that I believe is the root cause.
I don't think switching to sqlite would avoid this crash (if your stacktrace looks like the one above, starting and ending in prompt-toolkit), and there is not a convenient tool for migrating from json to sqlite anyway.

But you'll still be left with your large history. As a side quest, you might also try #3600, which gives you a `history gc --force` command to trim your history down to configured limits.  But review and update your history size limits first!  The default is not very big, and Xonsh has been ignoring it for a while.

**Comment by bhawkins** [comment]
I installed your PTK branch and so far so good. Will holler if I get another crash. Thanks!
